### DISCUSSION TITLE
Proposal: Scala macros

### DISCUSSION THREAD

**Discussion by darrenoakey** [dotnet/csharplang#3729] [Language Ideas]
Just want something exactly the same as Scala macros - that allows you to _change_ the ast at compile time.  For example, if  I could create a base class "Boring" such that if I did this:

```
class MyClass : Boring
{
    public readonly string a;
    public readonly string b;
}
```

it would

- generate a constructor that takes a and b
- and therefore compile without errors
- generate equals, testing and hash table

or I could create something like Cloneable

```
class Blah : Cloneable
{
     

}


```
which would add a clone method, which would rewrite the constructor and store any constructor parameters that aren't passed in, and then just call the constructor again with the same parameters

or I could create something like Locked

```
class Blah : Locked
{
     void Method1() {}
      string Method2() {return "hi"}
}
```

which would have an internal object _lockObject, and rewrite each of the methods as if I put lock (_lockObject) around the method body.

so far I haven't really found anything I wanted to do that I couldn't do in Scala macros - it somewhat obviates the need for a weaver.

**Comment by JoergWMittag** [comment]
While I am a huge fan of Scala, and think that more languages should steal from it, and I am a huge fan of macros, and think that more languages should have them, and I believe that Scala has *tremendously* advanced the state-of-the-art in statically-typed rich-syntax macros, the fact that macros were removed from the language, and re-introduced in an incompatible manner, in almost every major release over the last years, shows us that the state-of-the-art probably still needs some more advancing.

If you are advocating for "Scala macros", you at least should say *which ones*, since they have been re-designed multiple times.
